Date axis item

axil opened the pull request:
This is an improvement over PR #74 (see discussion there).

FilipeMaia commented:
Great job! I hope it gets merged...

2xB commented:
Thank you! This makes the custom DateAxis in `customPlot.py` obsolete, therefore I'd replace that with the new DateAxisItem. I've implemented that and sent a PR to @axil's fork.

This leaves the question open whether the examples should contain some kind of custom axis demonstration. Since I do not have a creative idea there, and the old custom axis example was probably anyways only given to provide a DateAxis implementation, I'd for now just place the DateAxisItem there. As soon as there is need for another custom axis item, that can then be added in the way it is needed.

axil commented:
@2xB I've fixed a bug in attachToPlotItem (it said that slot (3,1) was already taken), added two more examples (DateAxisItem itself and attachToPlotItem method to facilitate adding time axis to PlotWidget created with QtDesigner), merged your changes to the customPlot example and modified its description.

j9ac9k commented:
@2xB this ready to merge?  @axil thanks for the PR!

axil commented:
@2xB The new two examples do not follow the pattern in the other examples (only two lines behind the `if __name__ == __main__`). Is it important? Should I correct that?

2xB commented:
@axil If one could copy-paste the example into an interactive console without breaking it, that would be great. That is why the other examples are written like that. Also, while you are doing that, you could add the header that the other examples have:
```
"""
Demonstrates [...].

"""

import initExample ## Add path to library (just for examples; you do not need this)
```

That makes it possible to execute the example from the example folder without installing pyqtgraph.

As this turned into a collection of things to know for examples: anyways, here is one more thing: As soon as an example contains much information or is visually interesting, one should add it to the examples library in pyqtgraph/examples/utils.py , so it is also visible in the example browser ( [http://www.pyqtgraph.org/documentation/introduction.html#examples](http://www.pyqtgraph.org/documentation/introduction.html#examples) ).

2xB commented:
Also, I'd like to suggest one more patch that I will probably finish later today, moving the functionality of `attachToPlotItem` to `PlotItem`, probably refactoring some code from `PlotItem.__init__` into a seperate function, for the sake of more reusability and less maintained code.

2xB commented:
@axil The final thing that we would then need to do is to add documentation to the functions, that can be compiled into the API reference ( [http://www.pyqtgraph.org/documentation/apireference.html](http://www.pyqtgraph.org/documentation/apireference.html) ). For that we need to write down descriptions and all function parameters in a format similar to the descriptions given in e.g. `pyqtgraph/graphicsItems/PlotItem/PlotItem.py`. I don't know who of us wants to do that - probably you know more about the functions?

So I'd suggest you go along and work on the examples, I work on my patch, and you say how you'd like to approach documentation, and that's it!

2xB commented:
I've just seen that the changes needed for the documentation are very small, I will do them

axil commented:
@2xB Yes, attachToPlotItem is really hacky (both code and the name: not terribly intuitive would I attach to plot _item_ if I create a plot _widget_), I only included it because it allows to use QtDesigner without subclassing PlotWidget.

The notation

    pg.PlotWidget(axisItems = {'bottom': pg.DateAxisItem()}) 

still looks a bit heavyweight for a task as natural as making x axis accept timestamps. Would it be datetime instances it could be determined through typecheck with no args to PlotWidget at all (the way it is done in matplotlib), but unix timestamps are more memory-friendly than datetimes, so it's better to keep timestamps, yet something more concise could be more appropriate. Consider the bokeh library syntax:

    p = figure(x_axis_type='datetime')

Yet another alternative is to have a dedicated widget, say, DatetimePlotWidget (cf plot_date in matplotlib, but plot_date doesn't handle time at all, only dates).

Yes, I'd prefer that you updated the documentation. I've added the description to the examples and included the first of them to the `utils.py`.

2xB commented:
Great, thank you! Having a more natural way to set the axis types sounds good. I'd suggest not to do that before you reviewed my upcoming patch though, since that also changes part of the PlotItem __init__ function.

I've also seen that currently a user can crash the application just by zooming too far out. That is bad for practically every type of software based on pyqtgraph. My thought was just to automatically set min and max limits on the DateAxisItem. That is in theory done very easily:

```
class DateAxisItem(AxisItem):
    [...]
    def linkToView(self, view):
        super(DateAxisItem, self).linkToView(view)
        
        # Set default limits
        _min = ...
        _max = ...
        
        if self.orientation in ['right', 'left']:
            view.setLimits(yMin=_min, yMax=_max)
        else:
            view.setLimits(xMin=_min, xMax=_max)
```
Practically, I have no idea how to get values for minimum and maximum in this operating system dependent situation.

In the meantime, I am currently implementing a wrapper for utcfromtimestamp containing a try/except. But if there is a solution to the above concept, that'd be great!
Is there a way to get these minimum and maximum values?

2xB commented:
There is more to it than I thought: At least on my Linux machine utcfromtimestamp can output dates with years large enough that can't be handled by datetime. And if I wrap that with try/except, when zooming too far out, the axis scaling just stops working. Because of time constraints (pun not intended), I will therefore leave this issue for someone else to fix.

2xB commented:
So apart from what I've sent @axil as a Pull Request, I am currently seeing two things left:

- Crash when zooming too far out
- the heaviness of the notation to add a `DateAxisItem` ( https://github.com/pyqtgraph/pyqtgraph/pull/1154#issuecomment-611742614 )

I think we must fix the crash somehow, and be it with the code from my comment https://github.com/pyqtgraph/pyqtgraph/pull/1154#issuecomment-611792204 and very conservative limits, if nothing else works. The second point is optional, but it would of course great to have it implemented in `PlotItem`.

axil commented:
Thank you. Merged. I've renamed the second example accordingly. As for zoomout problem in bokeh they deal with it in the following way:

0) datetime

1) when the dates become too large they just display the years like integers:
![image](https://user-images.githubusercontent.com/170910/78992078-a85f8600-7b64-11ea-9a8f-bb587ba3813c.png)

2) after a certain threshold they stop displaying the ticks at all, allowing further zoomout:
![image](https://user-images.githubusercontent.com/170910/78992183-e0ff5f80-7b64-11ea-943c-24657ebb1ee8.png)

Both approaches look viable, second being slightly worse because it looks like a bug and makes it harder to zoomout x-axis (you have to point to a line and scroll the wheel instead of pointing to the rectangle with ticks as usual) but I'm not sure why they chose to have both approaches - one would suffice.

As a sidenote we could borrow zoom level thresholds from bokeh - they are fine-tuned a bit better better than current pyqtgraph implementation. Maybe it is windows-related problem though:
<img src="https://user-images.githubusercontent.com/170910/78992508-95998100-7b65-11ea-9304-ced2e9f35b11.png" width="500px">

Another idea for improvement would be to use current user locale when showing dates.

axil commented:
@2xB I've extended the working range to -1e13..1e13 years. After reaching this limit it hangs in `makeTicks` while loop. This can be dealt with but eventually it _will_ still get stuck somewhere (like max float value 1e308) and in my opinion 1e13 is already good enough as it is more than the age of the universe (~1.4e10 years). 

Yet a hard-coded limit somewhere about -1e12 .. 1e12 years would still be desireable to exclude the freezing/breaking completely.

j9ac9k commented:
I have no comments on the code, but I just wanted to thank you @axil for adding this functionality!  Also 👍 for referencing a bokeh implementation; love that library as well.

2xB commented:
Totally agreed!

- For the zoom level thresholds: Can the Bokeh thresholds just be implemented easily?
- For an easier way to set the axis: We could just use the same notation as Bokeh, which would make it easier to both set logarithmic scale and date scale. In that case, I could implement that, it should be pretty straight-forward.

2xB commented:
Since the second point is not strictly bound to DateAxisItem per se, I'd suggest to make it a seperate Pull Request after this PR is merged.

axil commented:
I gave it another try and found out that the case displayed on the screenshot seems to one of only two disurbing cases. All the other ones are ok. Can you reproduce in on linux? 
a) transition between 1-minutes and 30-seconds interval
![image](https://user-images.githubusercontent.com/170910/79134138-6ab66380-7dd7-11ea-9b6b-377c54d8f49b.png)
b) transition between 30-seconds and 15-seconds interval:
![image](https://user-images.githubusercontent.com/170910/79134237-91749a00-7dd7-11ea-9c50-29652ad2e7ad.png)
Is it possible to postpone this transition when zooming in a little bit so that the labels do not overlap?

j9ac9k commented:
> Is it possible to postpone this transition when zooming in a little bit so that the labels do not overlap?

If this transition is triggered by a signal being emitted, you could connect the signal via a Queued Connection:  https://doc.qt.io/qt-5/qt.html#ConnectionType-enum

axil commented:
By "postpone" I meant is to make this switch happen at the slightly more fine scale, not at the later moment in time.

2xB commented:
@axil I am not able to reproduce this on Linux using PyQt5.

axil commented:
Can you make a screenshot with 5-second zoomlevel? What is the width in pixels of the label like "13:31:45" there?

axil commented:
For me `self.maxTicksPerPt = 1/60.0` looks readable enough in this situation (like on the screenshots), but it is too sparse for other zoom levels.

2xB commented:
Here you go:
![image](https://user-images.githubusercontent.com/31772910/79147697-a5fb6700-7dc4-11ea-8ae2-2340b5b8277a.png)

Making the window slightly smaller:
![image](https://user-images.githubusercontent.com/31772910/79147738-b4498300-7dc4-11ea-8bf7-a1713951f2e8.png)

The label has a width of 49 pixels.

axil commented:
Ok, then the easiest solution would be to keep it constant, but to have a different constant on Windows: 64/49*60=78px. I tried setting it to 75px - it looks good enough already. 

Hmm. It might depend "font zoom level" OS setting. I'm currently working on 150% (which is a recommended setting, but it could vary with the screen resolution). I'd better check on other displays first.

2xB commented:
Could you run the following code and post the output? It prints the font and size used for the tick labels:

```python3
import time
from datetime import datetime, timedelta

import numpy as np
import pyqtgraph as pg
from pyqtgraph.Qt import QtGui

app = pg.mkQApp()

class FontInfoDateAxisItem(pg.DateAxisItem):
    def drawPicture(self, p, axisSpec, tickSpecs, textSpecs):
        pg.DateAxisItem.drawPicture(self, p, axisSpec, tickSpecs, textSpecs)
        
        font = p.font()
        print()
        print("pixelSize:", font.pixelSize())
        print("pointSize:", font.pointSize())
        print("family:", font.family())

# Create a plot with the date-time axis
w = pg.PlotWidget(axisItems = {'bottom': FontInfoDateAxisItem()})

# Plot some random data with timestamps in the last hour
now = time.time()
timestamps = np.linspace(now - 3600*24*365*1000, now, 100)
w.plot(x=timestamps, y=np.random.rand(100), symbol='o')

w.show()

## Start Qt event loop unless running in interactive mode or using pyside.
if __name__ == '__main__':
    import sys
    if (sys.flags.interactive != 1) or not hasattr(QtCore, 'PYQT_VERSION'):
        app.exec_()

```

For me it outputs 
```
pixelSize: -1
pointSize: 10
family: Noto Sans
```

axil commented:
pixelSize: -1
pointSize: 8
family: MS Shell Dlg 2

axil commented:
Yes, it is "font zoom level". On another monitor with this setting set to 100% everything looks fine with the default 60px size:
![image](https://user-images.githubusercontent.com/170910/79152288-1e2e5080-7df6-11ea-8ae7-4bdaecbf8da5.png)
The actual size is 43px here.

The code chunk you provided doesn't "feel" this OS setting:
pixelSize: -1
pointSize: 8
family: MS Shell Dlg 2

2xB commented:
```
physicalDotsPerInch: 164.76511156186612
logicalDotsPerInchX: 96.0
```

2xB commented:
I've extended the font info code to also output what Qt's FontMetrics assumes to be the size of the string `99:99:99` in pixels:

```python3
import time
from datetime import datetime, timedelta

import numpy as np
import pyqtgraph as pg
from pyqtgraph.Qt import QtGui


app = pg.mkQApp()

class FontInfoDateAxisItem(pg.DateAxisItem):
    def drawPicture(self, p, axisSpec, tickSpecs, textSpecs):
        pg.DateAxisItem.drawPicture(self, p, axisSpec, tickSpecs, textSpecs)
        
        font = p.font()
        print()
        print(p.fontMetrics().boundingRect("99:99:99"))
        print("pixelSize:", font.pixelSize())
        print("pointSize:", font.pointSize())
        print("family:", font.family())

# Create a plot with the date-time axis
w = pg.PlotWidget(axisItems = {'bottom': FontInfoDateAxisItem()})

# Plot some random data with timestamps in the last hour
now = time.time()
timestamps = np.linspace(now - 3600*24*365*1000, now, 100)
w.plot(x=timestamps, y=np.random.rand(100), symbol='o')

w.show()

## Start Qt event loop unless running in interactive mode or using pyside.
if __name__ == '__main__':
    import sys
    if (sys.flags.interactive != 1) or not hasattr(QtCore, 'PYQT_VERSION'):
        app.exec_()
```

For me it gives

```
PyQt5.QtCore.QRect(0, -14, 48, 18)
```
Can you give your output for the high DPI monitor?

axil commented:
The following works for me

        font_scale_factor = QtGui.QApplication.instance().primaryScreen().logicalDotsPerInchX() / 96
        self.maxTicksPerPt = 1/(60.0 * font_scale_factor)

@2xB Do you approve?

2xB commented:
@axil The issue is that this only works for the default text size. If one changes the font size, this breaks:
```python3
import time

import numpy as np
import pyqtgraph as pg
from pyqtgraph.Qt import QtGui

app = pg.mkQApp()

# Create a plot with the date-time axis
ax = pg.DateAxisItem()
f = QtGui.QFont()
f.setPointSize(20)
ax.setTickFont(f)
w = pg.PlotWidget(axisItems = {'bottom': ax})

# Plot some random data with timestamps in the last hour
now = time.time()
timestamps = np.linspace(now - 3600*24*365*1000, now, 100)
w.plot(x=timestamps, y=np.random.rand(100), symbol='o')

w.show()

## Start Qt event loop unless running in interactive mode or using pyside.
if __name__ == '__main__':
    import sys
    if (sys.flags.interactive != 1) or not hasattr(QtCore, 'PYQT_VERSION'):
        app.exec_()

```
Therefore I'd like to use the fontMetrics as seen above if possible.

axil commented:
Yes, agree, sounds reasonable. 

Also when dragged from highdpi to normal-dpi screen ideally we would handle the signal that the dpi has changed (something like QScreen.primaryScreenChanged). But it might be an overkill for such a nuance and we'd better wait for a bugreport with this issue :)

2xB commented:
@axil Can you run the code above and give the outputted `QtCore.QRect`? With that, I might quickly be able to make something.

2xB commented:
Sorry, missed that you already did that.

axil commented:
@2xB Using this branch in my work, I accidentally attached (through `setAxisItems`) one axis to two different plots and python died very ungracefully (good luck, Microsoft, figuring out what the problem is!). Is it possible to check if the axis has already been added to some plot prior to attaching to another?

2xB commented:
@axil What you are observing is that QGraphicsWidget might only be part of one layout in one scene. I've added a PR to your branch that should fix this by giving a meaningful error though.

j9ac9k commented:
Only the py2/qt4 builds are failing with the following errors:

```
=================================== FAILURES ===================================
____________________ testExamples[ customPlot.py - PyQt4 ] _____________________
[gw0] linux2 -- Python 2.7.17 /usr/share/miniconda/envs/test-environment-2.7/bin/python


Python 2.7.17 |Anaconda, Inc.| (default, Oct 21 2019, 19:04:46) 
[GCC 7.3.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> >>> ... ... ... ... ... ... ... ... ... ... ... ... ... ... ... ... Traceback (most recent call last):
  File "/home/vsts/work/1/s/pyqtgraph/graphicsItems/DateAxisItem.py", line 301, in paint
    self.fontScaleFactor = widget.window().windowHandle().screen().logicalDotsPerInchX() / 96
  File "/home/vsts/work/1/s/pyqtgraph/widgets/PlotWidget.py", line 80, in __getattr__
    raise AttributeError(attr)
AttributeError: windowHandle

Failed Plot Customization Example Test Located in customPlot.py 
```

Does Qt4 not have a `widget..window().windowHandle()`?

We're on the verge of killing off Qt4 support, so if this is going to be a significant amount of work to make Qt4 compatible, we can postpone this PR until after 0.11 is released.

j9ac9k commented:
Looks like there is a Qt4 issue:

```
=================================== FAILURES ===================================
____________________ testExamples[ customPlot.py - PyQt4 ] _____________________
[gw0] linux2 -- Python 2.7.17 /usr/share/miniconda/envs/test-environment-2.7/bin/python


Python 2.7.17 |Anaconda, Inc.| (default, Oct 21 2019, 19:04:46) 
[GCC 7.3.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> >>> ... ... ... ... ... ... ... ... ... ... ... ... ... ... ... ... Traceback (most recent call last):
  File "/home/vsts/work/1/s/pyqtgraph/graphicsItems/DateAxisItem.py", line 301, in paint
    self.fontScaleFactor = widget.window().windowHandle().screen().logicalDotsPerInchX() / 96
  File "/home/vsts/work/1/s/pyqtgraph/widgets/PlotWidget.py", line 80, in __getattr__
    raise AttributeError(attr)
AttributeError: windowHandle

Failed Plot Customization Example Test Located in customPlot.py 
```

It's the same failure on all py2/qt4 pipelines.  Does Qt4 not have a `.windowHandle()` attribute?

If this would be too much to get working for qt4, we can hold tight on this until after we release 0.11 (after which we will promptly kill off qt4 support).

j9ac9k commented:
hmm... but not an issue on windows qt4 stuff, that's so odd!

2xB commented:
The issue is indeed that the windowHandle() function is introduced in Qt 5.0, an oversight from me. I think as a possible workaround, one can assume a DPI of 96 for for for PyQt 4. Since with the last few commits we only use the resolution to scale padding around tick labels, one probably wouldn't even notice that workaround.

j9ac9k commented:
style check failure (why is it not showing up here?)

```
File has invalid line endings: ./examples/DateAxisItem.py
File has invalid line endings: ./examples/DateAxisItem_setAxisItems.py
```

axil commented:
@j9ac9k fixed the line endings

FilipeMaia commented:
Hurray!

j9ac9k commented:
I'll try and look at the code diff and play with the example later today.  Thanks everyone for your work, this is one of the first new features pyqtgraph has added in a very long time.  We've been on bug hunting mode for so long.

axil commented:
I've improved the second example so that it contained a complete Qt Designer project (like the one in designerExample.py). Now two dateaxis examples plot different functions: I'd prefer sin(1/x) from the first example over random data from the second, but in addition to dateaxis capabilities sin(1/x) also demonstrates #31, which might not be the best thing an example should do. 

I'm also missing microseconds a bit (currently 1 ms is the finest scale for the axis), but that is in no way urgent or indispensable.

j9ac9k commented:
Looking over the code diff, this is fantastic work.  I'm going to try and run the example code and see if I can break it; but looks like @2xB has been testing this pretty routinely. 

Thanks you again @axil !

axil commented:
I call off my idea with exponential years. It has little practical value but could lead to unexpected behaviour. For example, when you zoom out too far and then zoom in to year >9999 it won't show any ticks there. 

So I limited it to 1..9999 years, the range supported by python datetime library. Hopefully in 9999 the humanity will come up with something to resolve the 1K4 problem :)

It has an additional benefit that the years now have more natural width of "YYYY"  (instead of "-0.00000e00" so now the gaps between the year ticks will be smaller.

I feel like the branch is ready to merge now. Can anyone do some more testing?

2xB commented:
I've found one thing: We are using the order of a dict, namely `zoomLevels`. Dicts are only order preserving since 3.6: https://stackoverflow.com/a/39537308/8575607

axil commented:
> I've found one thing: We are using the order of a dict, namely zoomLevels. Dicts are only order preserving since 3.6: https://stackoverflow.com/a/39537308/8575607

This was correctly dealt with in some of the older versions.

j9ac9k commented:
> I've found one thing: We are using the order of a dict, namely `zoomLevels`. Dicts are only order preserving since 3.6: https://stackoverflow.com/a/39537308/8575607

Can either used an `OrderedDict` or keep using `dict` (but we will have to hold off on the merge until after 0.11 is released).

2xB commented:
Yes, it was my mistake and I will of course create a PR to fix this. I've just been in a call and just remembered that oversight, without remembering when and where it could happen, so I quickly wrote that message as a note. I am sorry for any confusion coming from both that issue and the very short comment.

What happened was that in PR https://github.com/axil/pyqtgraph/pull/6 I refactored `zoomLevels` to an array in an effort of not having to constantly recalculate the keys of `zoomLevels`, rewrote `setZoomLevelForDensity` because of both issues that we discussed in PR https://github.com/axil/pyqtgraph/pull/6 and the fact that somewhere the screen information I removed from the dict had to be considered (it is now hidden in `sizeOf`). To be able to use `break` in a loop over the `zoomLevels` array, I changed the order of the entries. Then I noticed that the non-screen-dependent part of the keys of the old `zoomLevels` dict could actually still be given as keys of a dictionary, re-adding the user-customization that was intended with that dict. Since for each `ZoomLevel` still the function `sizeOf` had to be calculated, I couldn't get rid of that for loop and go back to what was previously the implementation of `setZoomLevelForDensity`, so I left the new ordering and the new for loop, the latter one leading to the issue we are discussing here.

axil commented:
The @2xB's fix has been accepted. 

The branch is ready to be merged now!

j9ac9k commented:
@axil I'll give this a run through here in the next few hours.  Thanks!

j9ac9k commented:
Ran the example, looks great, one nit-pick.  The window generated by the example should have a title.

in `pyqtgraph/examples/DateAxisItem.py` before `w.show()`, call w.setTitle("pyqtgraph example: DateAxisItem')

axil commented:
@j9ac9k done

DanielRudnicki commented:
Hey, thanks a lot for that feature, it was awaited for a long time and finally there it is:)
Do you know when the merge may be?

axil commented:
@j9ac9k In my opinion this PR is mature enough to get merged to devel branch.

j9ac9k commented:
Hi, @axil ...I asked @ixjlyons and @campagnola to take a look at this PR as well 2 days ago, I wanted to give them at least to the end of the weekend, if I don't hear anything from them I'll merge a day from now.

axil commented:
@j9ac9k Sure, take your time. Just thought everybody has forgot about it.

j9ac9k commented:
Given the number of outstanding PRs, don't know why you would think that 🤣

axil commented on pyqtgraph/graphicsItems/DateAxisItem.py:
```diff
@@ -0,0 +1,319 @@
+from .AxisItem import AxisItem
+from ..pgcollections import OrderedDict
+
+__all__ = ['DateAxisItem', 'ZoomLevel']
```
I'd rather remove 'ZoomLevel' ;)

2xB replied:
Speaking of which: We advertise overriding `DateAxisItem.zoomLevels`, although the entries of that dict, such as `YEAR_MONTH_ZOOM_LEVEL`, are not available through `__all__`.

axil replied:
@2xB Does overriding `zoomLevels` have any practical purpose except for the locale representation of dates (for example someone might like "04.12" for the 4th of December instead of "4 Mon")? Also the names of the levels has become somewhat misleading: "HOUR_MINUTE_ZOOM_LEVEL" level stands for "day minute" actually and "HMS_ZOOM_LEVEL" stands for "second" in fact.

2xB replied:
The concept of a user-overridable `zoomLevels` dict actually comes from @3rdcycle in the PR #74 . I think the idea was that at that point, the zoom levels were not as precise as they are now, so there would be a good cause for the user to play around with the `zoomLevels` dict to make spacing better. Since we are now even DPI-aware, I would agree to you that there is probably no practical use to that and suggest to just remove this from the docstrings.

2xB commented on pyqtgraph/graphicsItems/AxisItem.py:
```diff
@@ -507,20 +507,29 @@ def linkedView(self):
+        
+        view.sigResized.connect(self.linkedViewChanged)
+        
+    def unlinkFromView(self):
```
Good catch! Since having a seperate method to unlink is the type of implementation you suggested over there, I'd suggest to just borrow the unit tests from over there. If that is agreed on, I'd do that, since `unlinkFromView` was from me.

axil commented:
Unicode issue:
![image](https://user-images.githubusercontent.com/170910/80372700-212d4480-88be-11ea-993f-89062c800b9e.png)
I've changed the locale to ru_RU:

```
import locale
locale.setlocale(locale.LC_TIME, "ru_RU")
```

I would expect something like "Янв", "Фев", ..., "Дек" in labels.

axil commented:
Ok, settings LC_ALL instead of LC_TIME fixes it. And not accepting '€' is the bug of strftime, not of pyqtgraph. So this one is resolved.

j9ac9k commented:
@axil Thanks for this PR, @2xB thanks for testing it along the way and thanks @ixjlyons for doing another passthrough.  This looks good to me, merging!

TheDigital1 commented:
Can someone provide an example, or better yet update the example code, to show how to adjust the axis tick scale? E.g. my use case I want to plot data this is over a day, so I would like a tick for every hour.
Thanks!

2xB commented:
@TheDigital1 The key feature of DateAxisItem is that it automatically adjusts the amount of ticks shown based on how far one has zoomed in (PyQtGraph is an interactive plotting library after all). Do I get it right that you want to overwrite this behavior manually to always show a fixed amount of ticks no matter how far the user zoomed in or out?
Btw. I would generally suggest you to open a new issue instead of posting on a merged Pull Request to assure that your question is actually seen by active people.

TheDigital1 commented:
@2xB Makes sense default functionality would be to automatically scale as you said for interactive reasons. However it just so happens in this one particular use case of mine I am using the plot in a non interactive way, and I would like the plot to look correct on the first initialization. The API reference talks about a "setZoomLevelForDensity(density)" function however I am not sure how to use this. 

**Some more details on my use case:**
For example the data I am plotting is 3 minute interval data over 12 hours:

Index Time (US/Pacific)  Total Instantaneous Real Power (kW)
0   2020-08-06 00:00:00                                   18
1   2020-08-06 00:03:00                                   27
2   2020-08-06 00:06:00                                   12
..                  ...                                  ...
238 2020-08-06 11:54:00                                   58
239 2020-08-06 11:57:00                                   89
240 2020-08-06 12:00:00                                   15

Upon initialization I would like the plot to show all data in view with grid lines at every hour. 

Thanks, I wasn't sure if making a new issue made sense as I figured it was just my lack of understanding of how it works and not really an issue. If you think this warrants a new issue let me know and I'll start one.

2xB commented:
There exists a "question" label that is put on issues that are not bugs in PyQtGraph, so don't worry there.

Some warning beforehand: The algorithm that you are going to deactivate is sensitive to the DPI of the monitor where the plot is displayed on and the size of the plot. So even if you always show the same plot makes you the responsible person for asserting that nothing overlaps. Also, if it is possible to draw ticks without overlapping labels for every hour, the algorithm will automatically do that for you. If it does not, probably the x axis will look relatively crowded.

Having said that, here is an example for you (a modified version of examples/customPlot.py):
```python3
# -*- coding: utf-8 -*-
import pyqtgraph as pg
import numpy as np

app = pg.mkQApp()

class CustomDateAxisItem(pg.DateAxisItem):
    def tickValues(self, minVal, maxVal, size):
        # Always use HOUR_MINUTE_ZOOM_LEVEL
        self.zoomLevel = pg.graphicsItems.DateAxisItem.DAY_HOUR_ZOOM_LEVEL

        # Generate values
        first_hour = np.ceil(minVal/3600)*3600
        ticks = np.arange(first_hour, maxVal, 3600)
        spacing = pg.graphicsItems.DateAxisItem.HOUR_SPACING
                # Has to be a spacing used for the self.zoomLevel we use
                # As a zoomLevel consists of two TickSpecs, the spacing indicates,
                # which of these two specs should be used to generate the ticks
        
        return [ (spacing, ticks) ]

vb = pg.ViewBox()

axis = CustomDateAxisItem(orientation='bottom')

pw = pg.PlotWidget(viewBox=vb, axisItems={'bottom': axis})

x = np.linspace(0.1,1.1,20) * (3600*12)
y = 1/x

pw.plot(x=x, y=y, symbol='o')
pw.show()

## Start Qt event loop unless running in interactive mode or using pyside.
if __name__ == '__main__':
    app.exec_()

```

2xB commented:
@TheDigital1 .

TheDigital1 commented:
@2xB  Just ran your example code. Looks good! I'll work to integrate it into my code and see if I get the results I'm looking for.
Thanks so much!
